README.md
<img src="https://user-images.githubusercontent.com/106826/43119494-78be9224-8ecb-11e8-9d1a-9fc6f3014b91.png" width="300" alt="API Sprout"/>

[![Go Report Card](https://goreportcard.com/badge/github.com/danielgtaylor/apisprout)](https://goreportcard.com/report/github.com/danielgtaylor/apisprout) [![Build Status](https://travis-ci.org/danielgtaylor/apisprout.svg?branch=master)](https://travis-ci.org/danielgtaylor/apisprout) [![GitHub tag (latest SemVer)](https://img.shields.io/github/tag/danielgtaylor/apisprout.svg)](https://github.com/danielgtaylor/apisprout/releases) [![Docker Pulls](https://img.shields.io/docker/pulls/danielgtaylor/apisprout.svg)](https://hub.docker.com/r/danielgtaylor/apisprout/)

A simple, quick, cross-platform API mock server that returns examples specified in an API description document. Features include:

- OpenAPI 3.x support
  - Uses operation `examples` or generates examples from `schema`
- Load from a URL or local file (auto reload with `--watch`)
- CORS headers enabled by default
- Accept header content negotiation
  - Example: `Accept: application/*`
- Prefer header to select response to test specific cases
  - Example: `Prefer: status=409`
- Server validation (enabled with `--validate-server`)
  - Validates scheme, hostname/port, and base path
  - Supports `localhost` out of the box
- Request parameter & body validation (enabled with `--validate-request`)
- Configuration via:
  - Files (`/etc/apisprout/config.json|yaml`)
  - Environment (prefixed with `SPROUT_`, e.g. `SPROUT_VALIDATE_SERVER`)
  - Commandline flags

Usage is simple:

```sh
# Load from a local file
apisprout my-api.yaml

# Validate server name and use base path
apisprout --validate-server my-api.yaml

# Load from a URL
apisprout https://raw.githubusercontent.com/OAI/OpenAPI-Specification/master/examples/v3.0/api-with-examples.yaml
```

## Docker Image

A hosted [API Sprout Docker image](https://hub.docker.com/r/danielgtaylor/apisprout/) is provided that makes it easy to deploy mock servers or run locally. For example:

```sh
docker pull danielgtaylor/apisprout
docker run -p 8000:8000 danielgtaylor/apisprout http://example.com/my-api.yaml
```

Configuration can be passed via environment variables, e.g. setting `SPROUT_VALIDATE_REQUEST=1`, or by passing commandline flags. It is also possible to use a local API description file via [Docker Volumes](https://docs.docker.com/storage/volumes/):

```
# Remember to put the full path to local archive YAML in -v
docker run -p 8000:8000 -v $FULLPATH/localfile.yaml:/api.yaml danielgtaylor/apisprout /api.yaml
```

## Installation

Download the appropriate binary from the [releases](https://github.com/danielgtaylor/apisprout/releases) page.

Alternatively, you can use `go get`:

```sh
go get github.com/danielgtaylor/apisprout
```

## Contributing

Contributions are very welcome. Please open a tracking issue or pull request and we can work to get things merged in.

## Release Process

The following describes the steps to make a new release of API Sprout.

1. Merge open PRs you want to release.
1. Select a new semver version number (major/minor/patch depending on changes).
1. Update `CHANGELOG.md` to describe changes.
1. Create a commit for the release.
1. Tag the commit with `git tag -a -m 'Tagging x.y.z release' vx.y.z`.
1. Build release binaries with `./release.sh`.
1. Push the commit and tags.
1. Upload the release binaries.

## License

Copyright &copy; 2018-2019 Daniel G. Taylor

http://dgt.mit-license.org/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
- Return header examples when possible.
- Update dependency versions.

## [1.3.0] - 2019-03-18
- Add `--add-server` to add a custom server when using `--validate-server`.
  This allows quickly adding a custom domain or base path that will properly
  validate.
- Add `--header` (short `-H`) option to specify a custom header when fetching
  the API document. This allows you to pass custom auth info.
- Add `readOnly` and `writeOnly` support to the example generator.
- Revamped support for `--validate-server` (short `-s`)
  - Requires the use of server base path(s) on the client.
  - Localhost is now always allowed on all known base paths.
  - Support for proxy headers (e.g. `X-Forwarded-Host`).
- Better support for resolving relative path references.
- Be more resilient to parser panics when using `--watch`
- Update Docker build to use Go 1.12 and Go modules.
- Enhance example-from-schema generation code. Support enums, string formats,
  array and object examples, min/max and min items.

## [1.2.0] - 2019-02-27
- Add support for reloading OpenAPI URLs via `/__reload` on the server.
- Support external references in OpenAPI loader.
- Update dependencies, simplify file loading.
- Support jsonapi.org content type (`application/vnd.api+json`).
- Switch from `dep` to Go modules.

## [1.1.1] - 2019-01-30
- Fix `OPTIONS` request to also include CORS headers.

## [1.1.0] - 2019-01-29
- Added the `--watch` (short `-w`) parameter to reload whenever the input file
  changes. This currently only works when using files on disk.
- Update Docker build to use Go 1.11.
- Generate examples from schema when no example is available.
- Fix path parameter validation.
- Add CORS headers. Disable with `--disable-cors`.
- Documentation updates.

## [1.0.1] - 2018-10-03
- Dependency updates, fixes string format validation bug.

## [1.0.0] - 2018-07-24
- Initial release.


example_test.go
package main

import (
	"encoding/json"
	"strings"
	"testing"

	"github.com/getkin/kin-openapi/openapi3"
	"github.com/stretchr/testify/assert"
)

var schemaTests = []struct {
	name string
	in   string
	out  string
}{
	// ----- Booleans -----
	{
		"Boolean",
		`{"type": "boolean"}`,
		`true`,
	},
	{
		"Boolean default",
		`{"type": "boolean", "default": false}`,
		`false`,
	},
	{
		"Boolean example",
		`{"type": "boolean", "example": false}`,
		`false`,
	},
	// ----- Numbers -----
	{
		"Integer",
		`{"type": "integer"}`,
		`0`,
	},
	{
		"Number",
		`{"type": "number"}`,
		`0.0`,
	},
	{
		"Number default",
		`{"type": "number", "default": 1.0}`,
		`1.0`,
	},
	{
		"Number example",
		`{"type": "number", "example": 2.0}`,
		`2.0`,
	},
	{
		"Number enum",
		`{"type": "number", "enum": [2.0, 4.0, 6.0]}`,
		`2.0`,
	},
	{
		"Number minimum",
		`{"type": "number", "minimum": 5.0}`,
		`5.0`,
	},
	{
		"Number exclusive minimum",
		`{"type": "number", "minimum": 5.0, "exclusiveMinimum": true}`,
		`6.0`,
	},
	{
		"Number maximum",
		`{"type": "number", "maximum": -1.0}`,
		`-1.0`,
	},
	{
		"Number exclusive maximum",
		`{"type": "number", "maximum": -1.0, "exclusiveMaximum": true}`,
		`-2.0`,
	},
	{
		"Number exclusinve minimum with max",
		`{"type": "number", "minimum": 5.0, "exclusiveMinimum": true, "maximum": 5.5}`,
		`5.25`,
	},
	{
		"Number exclusinve maximum with min",
		`{"type": "number", "minimum": -5.0, "maximum": -1.0, "exclusiveMaximum": true}`,
		`-3.0`,
	},
	{
		"Integer multiple of",
		`{"type": "integer", "minimum": 1, "multipleOf": 4}`,
		`4`,
	},
	// ----- Strings -----
	{
		"String",
		`{"type": "string"}`,
		`"string"`,
	},
	{
		"String default",
		`{"type": "string", "default": "def"}`,
		`"def"`,
	},
	{
		"String example",
		`{"type": "string", "example": "ex"}`,
		`"ex"`,
	},
	{
		"String enum",
		`{"type": "string", "enum": ["one", "two", "three"]}`,
		`"one"`,
	},
	{
		"String format date",
		`{"type": "string", "format": "date"}`,
		`"2018-07-23"`,
	},
	{
		"String format date-time",
		`{"type": "string", "format": "date-time"}`,
		`"2018-07-23T22:58:00-07:00"`,
	},
	{
		"String format time",
		`{"type": "string", "format": "time"}`,
		`"22:58:00-07:00"`,
	},
	{
		"String format email",
		`{"type": "string", "format": "email"}`,
		`"email@example.com"`,
	},
	{
		"String format hostname",
		`{"type": "string", "format": "hostname"}`,
		`"example.com"`,
	},
	{
		"String format ipv4",
		`{"type": "string", "format": "ipv4"}`,
		`"198.51.100.0"`,
	},
	{
		"String format ipv6",
		`{"type": "string", "format": "ipv6"}`,
		`"2001:0db8:85a3:0000:0000:8a2e:0370:7334"`,
	},
	{
		"String format uri",
		`{"type": "string", "format": "uri"}`,
		`"https://tools.ietf.org/html/rfc3986"`,
	},
	{
		"String format uri-template",
		`{"type": "string", "format": "uri-template"}`,
		`"http://example.com/dictionary/{term:1}/{term}"`,
	},
	{
		"String format json-pointer",
		`{"type": "string", "format": "json-pointer"}`,
		`"#/components/parameters/term"`,
	},
	{
		"String format regex",
		`{"type": "string", "format": "regex"}`,
		`"/^1?$|^(11+?)\\1+$/"`,
	},
	{
		"String format uuid",
		`{"type": "string", "format": "uuid"}`,
		`"f81d4fae-7dec-11d0-a765-00a0c91e6bf6"`,
	},
	{
		"String format password",
		`{"type": "string", "format": "password"}`,
		`"********"`,
	},
	{
		"String min length",
		`{"type": "string", "minLength": 10}`,
		`"stringstring"`,
	},
	{
		"String max length",
		`{"type": "string", "maxLength": 2}`,
		`"st"`,
	},
	{
		"String min & max length",
		`{"type": "string", "minLength": 8, "maxLength": 10}`,
		`"stringstri"`,
	},
	// ----- Arrays -----
	{
		"Array without items returns empty []",
		`{
			"type": "array"
		}`,
		`[]`,
	},
	{
		"Array of simple type",
		`{
			"type": "array",
			"items": {
				"type": "string"
			}
		}`,
		`["string"]`,
	},
	{
		"Array of simple type with example",
		`{
			"type": "array",
			"items": {
				"type": "string",
				"example": "I'm in an array"
			}
		}`,
		`["I'm in an array"]`,
	},
	{
		"Array of blank items fails",
		`{
			"type": "array",
			"items": {}
		}`,
		``,
	},
	{
		"Array with example",
		`{
			"type": "array",
			"example": [true, false, true],
			"items": {
				"type": "boolean"
			}
		}`,
		`[true, false, true]`,
	},
	{
		"Array of array of simple type",
		`{
			"type": "array",
			"items": {
				"type": "array",
				"items": {
					"type": "number"
				}
			}
		}`,
		`[[0]]`,
	},
	{
		"Array of objects",
		`{
			"type": "array",
			"items": {
				"type": "object",
				"required": ["foo", "bar"],
				"properties": {
					"foo": {
						"type": "boolean"
					},
					"bar": {
						"type": "string",
						"example": "baz"
					}
				}
			}
		}`,
		`[{"foo": true, "bar": "baz"}]`,
	},
	{
		"Array with min items (e.g. coordinates)",
		`{
			"type": "array",
			"minItems": 2,
			"items": {
				"type": "number"
			}
		}`,
		`[0, 0]`,
	},
	// ----- Objects -----
	{
		"Object without properties returns {}",
		`{
			"type": "object"
		}`,
		`{}`,
	},
	{
		"Object with example",
		`{
			"type": "object",
			"example": {
				"foo": 1
			},
			"properties": {
				"foo": {
					"type": "number"
				}
			}
		}`,
		`{"foo": 1}`,
	},
	{
		"Object with simple properties",
		`{
			"type": "object",
			"required": ["foo", "bar"],
			"properties": {
				"foo": {
					"type": "boolean"
				},
				"bar": {
					"type": "string",
					"example": "baz"
				}
			}
		}`,
		`{"foo": true, "bar": "baz"}`,
	},
	{
		"Object with complex properties",
		`{
			"type": "object",
			"properties": {
				"foo": {
					"type": "object",
					"properties": {
						"bar": {
							"type": "array",
							"items": {
								"type": "string"
							}
						}
					}
				}
			}
		}`,
		`{"foo": {"bar": ["string"]}}`,
	},
	{
		"Object with additional properties",
		`{
			"type": "object",
			"properties": {
				"foo": {
					"type": "number"
				}
			},
			"additionalProperties": {
				"type": "string"
			}
		}`,
		`{"foo": 0, "additionalPropertyName": "string"}`,
	},
	{
		"Object with additional properties error",
		`{
			"type": "object",
			"properties": {
				"foo": {
					"type": "number"
				}
			},
			"additionalProperties": {}
		}`,
		``,
	},
	// ----- Precedence -----
	{
		"Example before default",
		`{"type": "string", "default": "one", "example": "two"}`,
		`"two"`,
	},
	// ----- Modes -----
	{
		"Request mode",
		`{"type": "object", "required": ["normal", "readOnly", "writeOnly"],
			"properties": {
				"normal": {
					"type": "string"
				},
				"readOnly": {
					"type": "string",
					"readOnly": true
				},
				"writeOnly": {
					"type": "string",
					"writeOnly": true
				}
			}
		}`,
		`{"normal": "string", "writeOnly": "string"}`,
	},
	{
		"Response mode",
		`{"type": "object", "required": ["normal", "readOnly", "writeOnly"],
			"properties": {
				"normal": {
					"type": "string"
				},
				"readOnly": {
					"type": "string",
					"readOnly": true
				},
				"writeOnly": {
					"type": "string",
					"writeOnly": true
				}
			}
		}`,
		`{"normal": "string", "readOnly": "string"}`,
	},
}

func TestGenExample(t *testing.T) {
	for _, tt := range schemaTests {
		t.Run(tt.name, func(t *testing.T) {
			schema := &openapi3.Schema{}
			err := schema.UnmarshalJSON([]byte(tt.in))
			assert.NoError(t, err)
			m := ModeRequest
			if strings.Contains(tt.name, "Response") {
				m = ModeResponse
			}
			example, err := OpenAPIExample(m, schema)

			if tt.out == "" {
				// Expected to return an error.
				assert.Nil(t, example)
				assert.Error(t, err)
			} else {
				// Expected to match the output.
				var expected interface{}
				json.Unmarshal([]byte(tt.out), &expected)
				assert.EqualValues(t, expected, example)
			}
		})
	}
}


apisprout_test.go
package main

import (
	"testing"

	"github.com/getkin/kin-openapi/openapi3"
	"github.com/spf13/viper"
	"github.com/stretchr/testify/assert"
)

var localServerTests = []struct {
	name string
	in   []string
	out  []string
}{
	{
		"No servers",
		[]string{},
		[]string{},
	},
	{
		"Same path",
		[]string{
			"https://api.example.com/v1",
			"https://beta.api.example.com/v1",
		},
		[]string{
			"https://api.example.com/v1",
			"https://beta.api.example.com/v1",
			"http://localhost:8000/v1",
		},
	},
	{
		"Includes localhost already",
		[]string{
			"https://api.example.com/v1",
			"http://localhost:8000/v1",
		},
		[]string{
			"https://api.example.com/v1",
			"http://localhost:8000/v1",
		},
	},
	{
		"Invalid URL",
		[]string{
			"http://192.168.0.%31/",
		},
		[]string{},
	},
}

func TestAddLocalServers(t *testing.T) {
	viper.SetDefault("port", 8000)
	for _, tt := range localServerTests {
		t.Run(tt.name, func(t *testing.T) {
			servers := make([]*openapi3.Server, len(tt.in))
			for i, u := range tt.in {
				servers[i] = &openapi3.Server{
					URL: u,
				}
			}

			s := &openapi3.Swagger{
				Servers: servers,
			}

			err := addLocalServers(s)
			if len(tt.in) > 0 && len(tt.out) == 0 {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)

			results := make([]string, 0, len(tt.out))
			for _, server := range s.Servers {
				results = append(results, server.URL)
			}

			assert.Equal(t, tt.out, results)
		})
	}
}


example.go
package main

import (
	"fmt"

	"github.com/getkin/kin-openapi/openapi3"
)

// Mode defines a mode of operation for example generation.
type Mode int

const (
	// ModeRequest is for the request body (writes to the server)
	ModeRequest Mode = iota
	// ModeResponse is for the response body (reads from the server)
	ModeResponse
)

func getSchemaExample(schema *openapi3.Schema) (interface{}, bool) {
	if schema.Example != nil {
		return schema.Example, true
	}

	if schema.Default != nil {
		return schema.Default, true
	}

	if schema.Enum != nil && len(schema.Enum) > 0 {
		return schema.Enum[0], true
	}

	return nil, false
}

// stringFormatExample returns an example string based on the given format.
// http://json-schema.org/latest/json-schema-validation.html#rfc.section.7.3
func stringFormatExample(format string) string {
	switch format {
	case "date":
		// https://tools.ietf.org/html/rfc3339
		return "2018-07-23"
	case "date-time":
		// This is the date/time of API Sprout's first commit! :-)
		return "2018-07-23T22:58:00-07:00"
	case "time":
		return "22:58:00-07:00"
	case "email":
		return "email@example.com"
	case "hostname":
		// https://tools.ietf.org/html/rfc2606#page-2
		return "example.com"
	case "ipv4":
		// https://tools.ietf.org/html/rfc5737
		return "198.51.100.0"
	case "ipv6":
		// https://tools.ietf.org/html/rfc3849
		return "2001:0db8:85a3:0000:0000:8a2e:0370:7334"
	case "uri":
		return "https://tools.ietf.org/html/rfc3986"
	case "uri-template":
		// https://tools.ietf.org/html/rfc6570
		return "http://example.com/dictionary/{term:1}/{term}"
	case "json-pointer":
		// https://tools.ietf.org/html/rfc6901
		return "#/components/parameters/term"
	case "regex":
		// https://stackoverflow.com/q/3296050/164268
		return "/^1?$|^(11+?)\\1+$/"
	case "uuid":
		// https://www.ietf.org/rfc/rfc4122.txt
		return "f81d4fae-7dec-11d0-a765-00a0c91e6bf6"
	case "password":
		return "********"
	}

	return ""
}

// excludeFromMode will exclude a schema if the mode is request and the schema
// is read-only, or if the mode is response and the schema is write only.
func excludeFromMode(mode Mode, schema *openapi3.Schema) bool {
	if schema == nil {
		return true
	}

	if mode == ModeRequest && schema.ReadOnly {
		return true
	} else if mode == ModeResponse && schema.WriteOnly {
		return true
	}

	return false
}

// OpenAPIExample creates an example structure from an OpenAPI 3 schema
// object, which is an extended subset of JSON Schema.
// https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#schemaObject
func OpenAPIExample(mode Mode, schema *openapi3.Schema) (interface{}, error) {
	if ex, ok := getSchemaExample(schema); ok {
		return ex, nil
	}

	switch {
	case schema.Type == "boolean":
		return true, nil
	case schema.Type == "number", schema.Type == "integer":
		value := 0.0

		if schema.Min != nil && *schema.Min > value {
			value = *schema.Min
			if schema.ExclusiveMin {
				if schema.Max != nil {
					// Make the value half way.
					value = (*schema.Min + *schema.Max) / 2.0
				} else {
					value++
				}
			}
		}

		if schema.Max != nil && *schema.Max < value {
			value = *schema.Max
			if schema.ExclusiveMax {
				if schema.Min != nil {
					// Make the value half way.
					value = (*schema.Min + *schema.Max) / 2.0
				} else {
					value--
				}
			}
		}

		if schema.MultipleOf != nil && int(value)%int(*schema.MultipleOf) != 0 {
			value += float64(int(*schema.MultipleOf) - (int(value) % int(*schema.MultipleOf)))
		}

		if schema.Type == "integer" {
			return int(value), nil
		}

		return value, nil
	case schema.Type == "string":
		if ex := stringFormatExample(schema.Format); ex != "" {
			return ex, nil
		}

		example := "string"

		for schema.MinLength > uint64(len(example)) {
			example += example
		}

		if schema.MaxLength != nil && *schema.MaxLength < uint64(len(example)) {
			example = example[:*schema.MaxLength]
		}

		return example, nil
	case schema.Type == "array", schema.Items != nil:
		example := []interface{}{}

		if schema.Items != nil && schema.Items.Value != nil {
			ex, err := OpenAPIExample(mode, schema.Items.Value)
			if err != nil {
				return nil, fmt.Errorf("can't get example for array item")
			}

			example = append(example, ex)

			for uint64(len(example)) < schema.MinItems {
				example = append(example, ex)
			}
		}

		return example, nil
	case schema.Type == "object", len(schema.Properties) > 0:
		example := map[string]interface{}{}

		for k, v := range schema.Properties {
			if excludeFromMode(mode, v.Value) {
				continue
			}

			ex, err := OpenAPIExample(mode, v.Value)
			if err != nil {
				return nil, fmt.Errorf("can't get example for '%s'", k)
			}

			example[k] = ex
		}

		if schema.AdditionalProperties != nil && schema.AdditionalProperties.Value != nil {
			addl := schema.AdditionalProperties.Value

			if !excludeFromMode(mode, addl) {
				ex, err := OpenAPIExample(mode, addl)
				if err != nil {
					return nil, fmt.Errorf("can't get example for additional properties")
				}

				example["additionalPropertyName"] = ex
			}
		}

		return example, nil
	}

	return nil, ErrNoExample
}
